### DISCUSSION TITLE
Setting Request Headers in some useQuery requests react nextjs  🥺

### DISCUSSION THREAD

**Discussion by okaforcj** [trpc/trpc#1686] [Q&A] [answered]
Hello. Pretty please. I am using firebase authentication. I want to set the request headers dynamically for some queries in react using the request headers. ATM, i dont know how to add request headers to this particular request becuase the token is gotten dynamically and is only specific on a particular request. Can anyone shed some light on this or have experience doing something similar?

``` js script
const { data, refetch: getUser } = trpc.useQuery(
    ["users.find", uidRef.current ?? ""],
    {
      enabled: Boolean(user),
    }
  );

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, async (userCred) => {
      if (userCred) {
        const uid = userCred.uid;
        const token = await userCred.getIdToken();
        await getUser()
          .then((user) => {
            if (user.data) {
              console.log(user.data);
              setUser(user.data);
            }
          })
          .catch((err) => {
            //todo: toastify error
            console.error(err);
            logOut();
          });
        // ...
      } else {
        setUser(null);
        // User is signed out
        // ...
      }
    });
    return () => unsubscribe();
  }, []);
```

**Comment by KATT** [comment]
You can store the auth obj in a cookie/local storage/on the `window` object and use it in the `headers() {}`-callback where the client is created

Examples:

- https://github.com/trpc/trpc/blob/a581fb20589f376a76233bae1670d62f4e9e0367/examples/express-server/src/client.ts#L55-L60
- https://trpc.io/docs/react#3-add-trpc-providers

**Reply by KATT** [reply]
Some more mock code you might wanna get inespired by:

```tsx
const [token, setToken] = useState(() => window.localStorage.get('token'))
const trpc = trpc.useContext()
useEffect(() => {
  trpc.invalidateQueries() // invalidate all queries on auth change
  window.localStorage.set('token', token);
}, [token])

// [..] do auth stuff from firebase and use `setToken()`

```

**Reply by KATT** [reply]
If this helps, I would love it if you could make a PR and update the docs in an appropriate location 🙏

**Reply by okaforcj** [reply]
Thanks for replying!!! I did infact do something similar but forgot to invalidate all queries 🤦🏾‍♂️ opted for session storage just cos i'm scared of persisting tokens that live forever lol but thanks again for this, I kinda wanted re-assurance that i was following what was intended

``` js script
useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, async (userCred) => {
      if (userCred) {
        sessionStorage.setItem("token", await userCred.getIdToken());
        await getUser()
          .then((user) => {
            if (user.data) {
              setUser(user.data);
            } else {
              console.log("no data");
            }
          })
          .catch((err) => {
            //todo: toastify error
            console.error(err);
            logOut();
          });
        // ...
      } else {
        setUser(null);
        sessionStorage.removeItem("token");
      }
    });
    return () => unsubscribe();
  }, []);
```

and 

``` js script
return {
      url,
      headers() {
        return {
          Authorization: sessionStorage.getItem("token") ?? "",
        };
      },
    }
```

**Reply by okaforcj** [reply]
Also @KATT I saw your repo for the first time on fireship.io and ive spent the past 2 days refactoring my companies nextjs api routes to trpc. Best repo (hidden gem for typed apis) by a longggg way. I really love this project, thanks for spending your time with it. I will put up a PR for doc changes and any other help I can bring, I will!

**Comment by dandin84** [comment, marked as answer]
When I am using only the firebase auth without reaching the firestore, the token do not refresh automatically before a hour passed. 
Would you consider as best practice to use the firebase auth.currentUser?.getIdToken() instead of the sessionStorage approach?

1. Handle auth change in a context:
src/context/auth.context.tsx
```javascript
export const AuthProvider: FC<IAuthProvider> = ({ children }) => {
  const [session, setSession] = useState<IAuthContext>(init);

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, async (user) => {
      if (user) {
          setSession({ loading: false, user: user });
      } else {
        await signOut(auth);
        setSession({ ...init, loading: false });
      }
    });
    return unsubscribe;
  }, []);

  return (
    <AuthContext.Provider value={session}>{children}</AuthContext.Provider>
  );
};
```

2. Wrap the app with the auth provider:
src/pages/_app.tsx
```javascript
const MyApp: AppType = ({ Component, pageProps }) => {
  return (
    <AuthProvider>
      <Component {...pageProps} />
    </AuthProvider>
  );
};

export default trpc.withTRPC(MyApp);
```

3. Send the token via headers using firebase auth.currentUser:
src/utils/trpc.ts
```javascript
import { auth } from "../utils/firebase.client";

export const trpc = createTRPCNext<AppRouter>({
  config() {
    return {
      transformer: superjson,
      links: [
        loggerLink({
          enabled: (opts) =>
            process.env.NODE_ENV === "development" ||
            (opts.direction === "down" && opts.result instanceof Error),
        }),
        httpBatchLink({
          url: `${getBaseUrl()}/api/trpc`,
          /**
           * Headers will be called on each request.
           */
          async headers() {
            const token = (await auth.currentUser?.getIdToken()) || "";
            return token
              ? {
                  Authorization: `Bearer ${token}`,
                }
              : {};
          },
        }),
      ],
    };
  },
  ssr: false,
});
```

4. Verify the token in trpc context:
src/server/trpc/context.ts
```javascript
export const createContext = async (
  opts: trpcNext.CreateNextContextOptions
) => {
  // This is just an example of something you'd might want to do in your ctx fn
  async function getUserFromHeader() {
    const authorization = opts.req.headers.authorization;
    if (!authorization) return null;

    const token = authorization.split(" ")[1];
    if (!token) return null;

    const decodedIdToken = await auth.verifyIdToken(token);
    if (!decodedIdToken) return null;

    return decodedIdToken;
  }

  const session = await getUserFromHeader();

  return await createContextInner({
    session,
  });
};
```

5. Create auth procedure:
src/server/trpc/trpc.ts
```javascript
export const authedProcedure = t.procedure.use(({ ctx, next }) => {
  if (!ctx.session || !ctx.session.uid) {
    throw new TRPCError({ code: "UNAUTHORIZED" });
  }
  return next();
});
```

6. Use the auth procedure:
src/server/trpc/router/example.ts
```javascript
export const exampleRouter = t.router({
  hello: t.procedure
    .input(z.object({ text: z.string().nullish() }).nullish())
    .query(({ input, ctx }) => {
      return {
        greeting: `Hello ${input?.text ?? "world"}`,
      };
    }),
  test: authedProcedure
    .input(z.object({ text: z.string() }))
    .query(({ input, ctx }) => {
      return {
        greeting: `Hello ${input.text}`,
      };
    }),
});
```

**Reply by okaforcj** [reply]
> When I am using only the firebase auth without reaching the firestore, the token do not refresh automatically before a hour passed. Would you consider as best practice to use the firebase auth.currentUser?.getIdToken() instead of the sessionStorage approach?

@dandin84  you are absolutely right, 2months ago i faced the exact issue and fixed it with what youve suggested. That session solution would work potentially if you were managing refresh tokens yourself and keeping the session storage up to date, but I, like you, was using firebase and it automatically refreshes when the token expires provided the user is still logged in. Session storage cant be considered a better approach because it is even wrong and might cause issues for people. Apologies for not doing my due diligence and updating on here and thanks for putting it back into the discussion

**Reply by dandin84** [reply]
@okaforcj This is great, thanks for your answer :pray:

**Reply by lewiscasewell** [reply]
Hi this is great is there a way you would similarly do this on React Native?

**Reply by okaforcj** [reply]
dont see whats different with react-native, its pretty straight forward,

This is one of the configs i had for an RN project in the past

``` js script
export default function App() {
  const [queryClient] = useState(
    () =>
      new QueryClient({
        defaultOptions: {
          queries: {
            refetchOnWindowFocus: true,
          },
        },
      })
  )

  const wsClient = createWSClient({
    url: socketUri,
  })

  const [trpcClient] = useState(() =>
    trpc.createClient({
      transformer: superjson,

      links: [
        splitLink({
          condition(op) {
            return op.type == 'subscription'
          },
          // when condition is false, use normal request
          false: httpLink({
            url: `${serverUri}/trpc`,
          }),
          true: wsLink({
            client: wsClient,
          }),
        }),
      ],

      async headers() {
        const tokenValue = await auth.currentUser?.getIdToken()
        return {
          authorization: tokenValue ?? '',
        }
      },
    })
  )

  return (
    <trpc.Provider client={trpcClient} queryClient={queryClient}>
      <QueryClientProvider client={queryClient}>
        <ActionSheetProvider>
          <NativeBaseProvider>
            <PortalProvider>
              <Providers />
            </PortalProvider>
          </NativeBaseProvider>
        </ActionSheetProvider>
      </QueryClientProvider>
    </trpc.Provider>
  )
}
```

**Reply by u007** [reply]
hi, is this trpc.Provider based on latest trpc 10? my version 10 states trpc.Provider is undefined.
anyone have any clue? im using createTRPCNext<AppRouter>, how do i create a provider like in the code?

**Comment by dandin84** [comment]
Hi @okaforcj, 

In relate to the code I posted earlier: https://github.com/trpc/trpc/discussions/1686#discussioncomment-3831581

I was expecting by using this auth procedure: 
```
export const authedProcedure = t.procedure.use(({ ctx, next }) => {
  if (!ctx.session || !ctx.session.uid) {
    throw new TRPCError({ code: "UNAUTHORIZED" });
  }
  return next();
});
```
That at the router function the session will for sure not be null and have the uid:
<img width="530" alt="image" src="https://user-images.githubusercontent.com/6383184/196745973-d784d00d-5b43-4029-910f-7605f4e8256e.png">


But apparently the typescript yell so I must to add "?" and check if the userId exist:
<img width="539" alt="image" src="https://user-images.githubusercontent.com/6383184/196745590-4f455695-919f-43df-acea-5919338d662e.png">

So basically it seems that the typescript rely only on the type that I declared in the context:
<img width="586" alt="image" src="https://user-images.githubusercontent.com/6383184/196747049-c25e5f64-7073-4e91-8207-b85bee6962c8.png">

But not on the auth validation in the auth procedure...

I am wondering, is it make sense or am I doing something wrong and you would recommend a better approach?

**Reply by okaforcj** [reply]
are you sure youre applying the context types correctly 

see sample context types i created for one of. my previous projects, hope it helps`


``` js script
export const createExpressContext = ({
  req,
  res,
}: trpcExpress.CreateExpressContextOptions) => ({
  token: req.headers.authorization,
  socket: req.socket,
  uid: null,
  origin: req.headers.origin,
});
//sorthis outs
export const createWSContext = ({ req, res }: any) => ({
  token: req.headers.authorization,
  socket: req.socket,
  uid: null,
  origin: req.headers.origin,
});
export type WSContext = inferAsyncReturnType<typeof createWSContext>;
export type ExpressContext = inferAsyncReturnType<typeof createExpressContext>;

export const trpcRouter = () => {
  return trpc.router<WSContext | ExpressContext>();
};

```

**Reply by dandin84** [reply]
@okaforcj 

Thanks for your reply!
I am not sure how much my code aligned with yours but you made me to understand that i am missing something so I checked the docs more carefully and I found out that I missed the infer part in the return of the auth procedure:
https://trpc.io/docs/v10/context

<img width="748" alt="image" src="https://user-images.githubusercontent.com/6383184/196797843-db6dadfc-16f1-469a-b173-d4e30b043b09.png">

Now its works correctly, thanks :pray:
